jwst/lib/pipe_utils.py

```diff
@@ -76,3 +83,69 @@ def is_irs2(model):
         return True
     else:
         return False
+
+
+def match_nans_and_flags(input_model):
+    """Ensure data, error, variance, and DQ are marked consistently for invalid data.
+
+    Invalid data is assumed to be any pixel set to NaN in any one of the
+    data, error, or variance arrays, or else set to the DO_NOT_USE flag
+    in the DQ array.
+
+    The input model is updated in place with NaNs or DO_NOT_USE flags, as
+    appropriate, at all invalid data locations.
+
+    Parameters
+    ----------
+    input_model : DataModel
+        Input model containing some combination of data, dq, err, var_rnoise,
+        var_poisson, and var_flat extensions. These extensions must all have
+        matching dimensions if present.
+    """
+    # Build up the invalid data flags from each available data extension.
+    is_invalid = None
+    data_shape = None
+    nan_extensions = ['data', 'err', 'var_rnoise', 'var_poisson', 'var_flat']
+    for extension in nan_extensions:
+        if not hasattr(input_model, extension):
+            continue
+        data = getattr(input_model, extension)
+        if is_invalid is None:
+            is_invalid = np.isnan(data)
+            data_shape = data.shape
+        else:
+            if data.shape != data_shape:
+                log.warning(f"Mismatched data shapes; skipping invalid data "
+                            f"updates for extension '{extension}'")
+                continue
+            is_invalid |= np.isnan(data)
+
+    # Add in invalid flags from the DQ extension if present
+    if hasattr(input_model, 'dq'):
+        do_not_use = (input_model.dq & dqflags.pixel['DO_NOT_USE']).astype(bool)
+        if is_invalid is None:
+            is_invalid = do_not_use
+            data_shape = input_model.dq.shape
+        else:
+            if input_model.dq.shape != data_shape:
+                log.warning("Mismatched data shapes; skipping invalid data "
+                            "updates for extension 'dq'")
+            else:
+                is_invalid |= do_not_use
+
+    # Nothing to do if no extensions were found
```

Comment:
In what case would you want to pass a DataModel through here that has none of the extensions listed in line 108?

Reply:
You wouldn't, really.  I just wanted it to make sure it was as general as possible, so it could be called everywhere.